Python/compile.c

```diff
@@ -325,6 +325,14 @@ enum {
     COMPILER_SCOPE_COMPREHENSION,
 };
 
+typedef struct cfg_ {
+    basicblock *entryblock; /* Where control flow begins */
+    /* Pointer to the most recently allocated block.  By following
+       b_list links, you can reach all allocated blocks. */
+    basicblock *block_list;
+    basicblock *curblock; /* pointer to current block */
```

Comment:
Wouldn't `curblock` be better in the compiler?
Concept of a "current" block makes sense to the front-end, but does it make sense to the optimizer or assembler?

Reply:
We use it in cfg_addop to decide whether we need to implicitly start a new block. So we would need the caller (compiler or unit test or whoever is constructing a cfg) to pass it in.

Reply:
That implies (to me) that `addop` should also belong to the compiler.

Reply:
The unit test doesn’t have a compiler.

Reply:
What are you testing?
Adding an instruction to a CFG doesn't make sense as an operation (at least not to me).

Reply:
Testing the optimizations. The test needs to construct a cfg and feed it into the cfg_optimize function. Then check that the output cfg is what is expected.

Reply:
How do you propose to create the CFG?

It seems to me that the concept of "current block" is useful when creating a CFG, but meaningless when optimizing it and assembling it.

Reply:
> How do you propose to create the CFG?

I have a draft here: https://github.com/python/cpython/compare/main...iritkatriel:cpython:optimize_func1?expand=1#diff-687f4a736f1f6ad4b6cd4ff65c8e0113408304e8d65530dbd41b3f59a081a126R568

See instructions_to_cfg in Modules/_testinternalcapi.c

 
> It seems to me that the concept of "current block" is useful when creating a CFG, but meaningless when optimizing it and assembling it.

I don't think it's so clear cut. There are places during optimization/assembly where need to add instructions, or even a whole new block, to the CFG.

Reply:
> I have a draft here

I like having the ability to convert a list of instructions into a code object. It would save us this sort of hackery 
https://github.com/python/cpython/pull/93221/files#diff-22e653d878778ca28317261f7826545288d02fd9980cf485d0329f28a34993f2R2060
but I feel it belongs in the compiler, not in testcapi.

> There are places during optimization/assembly where need to add instructions, or even a whole new block, to the CFG.

But we aren't adding them to the CFG as a whole, but a particular place in the CFG, namely a block.

Reply:
> But we aren't adding them to the CFG as a whole, but a particular place in the CFG, namely a block.

If you look in this PR at the two places where, during optimization, I pass g rather than g->entryblock, you will see how it's done now. Previously we seemed to care where in the b_list the new block was, but that doesn't matter - the only things we use b_list for now is deallocating the list and counting how many blocks there are. What really matters is where the new block is in the b_next list. 

I'm actually not sure why we bother with the b_list links, we could just use a PyArena for the basicblocks of a CFG (and add a couter of allocations so we can blow up if it gets to more than `SIZE_MAX/sizeof(basicblock*)` )

Reply:
> I like having the ability to convert a list of instructions into a code object. It would save us this sort of hackery
> https://github.com/python/cpython/pull/93221/files#diff-22e653d878778ca28317261f7826545288d02fd9980cf485d0329f28a34993f2R2060
> but I feel it belongs in the compiler, not in testcapi.

We can move the code back to compile.c if it's useful for things other than tests. But note that for tests I don't want something that constructs a code object. I want to construct a CFG, run the optimizer on it and give me back the optimised CFG.

Reply:
As discussed offline - I’ve renamed cfg to cfg_builder and it’s entry block field to cfg. This indicated more clearly what the new struct is for.